File: common/src/client/java/juuxel/adorn/client/FluidRenderingBridge.java
package juuxel.adorn.client;

import juuxel.adorn.fluid.FluidReference;
import juuxel.adorn.util.InlineServices;
import juuxel.adorn.util.Services;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.block.BlockAndTintGetter;
import net.minecraft.client.renderer.block.FluidModel;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.TooltipFlag;
import net.minecraft.world.level.material.FluidState;
import org.jspecify.annotations.Nullable;

import java.util.List;

@InlineServices
public interface FluidRenderingBridge {
    default FluidModel getFluidModel(FluidReference volume) {
        var mc = Minecraft.getInstance();
        FluidState fluidState = volume.getFluid().defaultFluidState();
        return mc.getModelManager().getFluidStateModelSet().get(fluidState);
    }

    default @Nullable TextureAtlasSprite getStillSprite(FluidReference volume) {
        return getFluidModel(volume).stillMaterial().sprite();
    }

    int getColor(FluidReference volume, @Nullable BlockAndTintGetter world, @Nullable BlockPos pos);

    default int getColor(FluidReference volume) {
        return getColor(volume, null, null);
    }

    boolean fillsFromTop(FluidReference volume);

    List<Component> getTooltip(FluidReference volume, TooltipFlag type, @Nullable Integer maxAmountInLitres);

    @InlineServices.Getter
    static FluidRenderingBridge get() {
        return Services.load(FluidRenderingBridge.class);
    }
}

File: common/src/main/kotlin/juuxel/adorn/platform/FluidRenderingBridge.kt
package juuxel.adorn.platform

import juuxel.adorn.fluid.FluidReference
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.item.TooltipContext
import net.minecraft.client.texture.Sprite
import net.minecraft.text.Text
import net.minecraft.util.math.BlockPos
import net.minecraft.world.BlockRenderView

// TODO: NO NO NO, ENVIRONMENT SHOULD NOT BE HERE!!!
interface FluidRenderingBridge {
    val bucketVolume: Long

    @Environment(EnvType.CLIENT)
    fun getStillSprite(volume: FluidReference): Sprite?

    @Environment(EnvType.CLIENT)
    fun getColor(volume: FluidReference, world: BlockRenderView? = null, pos: BlockPos? = null): Int

    @Environment(EnvType.CLIENT)
    fun fillsFromTop(volume: FluidReference): Boolean

    @Environment(EnvType.CLIENT)
    fun getTooltip(volume: FluidReference, context: TooltipContext, maxAmountInLitres: Int?): List<Text>

    companion object {
        fun get() = PlatformBridges.get().fluidRendering
    }
}
